_posts/2018-03-28-building-coreboot-on-minnowboard.md

```diff
@@ -0,0 +1,244 @@
+version of coreboot for Bay Trail however it won't work without MCU code which
+is confidential (available for trusted vendors) so we won't cover this option.
+
+Intel sites redirect to https://github.com/IntelFsp/FSP (branch BayTrail) where
+you can find it. We are interested mostly in BayTrailFspBinPkg/FspBin/BayTrailFSP.fd
+which is file we have to provide to coreboot. FSP can be configured using BCT
+(Binary Configuration Tools), which is optional — needed to enable Secure boot
+and FastBoot. The tool is available only as Windows application, but works well
+using Wine too.
+
+In the same package you can find VgaBIOS which can be used too if you want to
+use graphic card: BayTrailFspBinPkg/Vbios/Vga.dat
+
+ME region
+---------
+
+The simplest way to prevent ME region corruption is to read ROM layout from
+original firmware image. In `utils/ifdtool` of coreboot source tree we can
+find program for reading layout from ROM image. The image can be taken from
+firmware package or read using flashrom, using SPI interface:
+
+```sh
+sudo flashrom -pdediprog -r minnow.rom
+```
+
+Note that you may need to adjust `-p` option according to used SPI programmer.
+Use `-pinternal` if you use MinnowBoard's internal programmer.
+
+```sh
+~/code/coreboot> cd util/ifdtool/
+~/code/coreboot/util/ifdtool> make
+~/code/coreboot/util/ifdtool> ./ifdtool -f ../../mb.layout minnow.rom
+File minnow.bin is 8388608 bytes
+Wrote layout to ../../mb.layout
+~/code/coreboot/util/ifdtool> cat ../../mb.layout
+00000000:00000fff fd
+00400000:007fffff bios
+00001000:003fffff me
+00000000:00000fff gbe
+```
+
+While flashing coreboot we should inform flashrom only to write `bios` region
+
+```sh
+sudo flashrom -l mb.layout -ibios -pdediprog -w build/coreboot.rom
+```
+
+Toolchain
+---------
+
+coreboot needs specific versions of toolchain components, we have to take care
+of this. Makefile has rule for building toolchain:
+
+```sh
+# make sure, we have all dependencies:
+apt-get install git build-essential gnat flex bison libncurses5-dev wget zlib1g-dev
+
+make crossgcc-i386
+```
+
+Despite the fact that that we usualy want to boot 64-bit OS, coreboot code is
+compiled for 32-bit. Moreover we have noticed that there's problem with libpayload
+on 64-bit toolchain.
+
+However this procedure may be problematic. It may take much time on older machine
+and compilation may fail on some GCC versions. It's much easier to use docker,
+there is dedicated docker image with toolchain for coreboot:
+
+```
+sudo apt-get install docker #if needed, install
+sudo systemctl start docker # run docker service if needed
+sudo docker pull coreboot/coreboot-sdk:1.50
+```
+
+There is also our image with additional FSP package (under default path):
+`3mdeb/coreboot-trainings-sdk:latest`.
+
+To run shell in docker environment:
+
+```sh
+docker run -u root -it -v $PWD:/home/coreboot/coreboot --rm  \
+    coreboot/coreboot-sdk:1.50 /bin/bash
+```
```

Comment:
@BartekLew there is silent assumption that we are in coreboot directory.

Reply:
:+1: